src/app/modules/submission/encrypt-submission/encrypt-submission.utils.ts

```diff
@@ -186,3 +190,34 @@ export const getStripePaymentMethod = (
     },
   }
 }
+
+/**
+ * Sanitizes the payment fields from the form and the incoming submission
+ * The payment products from incoming submission can be freely altered by the respondent
+ * which could result in undesirable data seeded into our database
+ * @param form
+ * @param dirtyPaymentProducts
+ */
+export const sanitisePaymentProducts = (
+  form: IPopulatedEncryptedForm,
+  dirtyPaymentProducts: ProductItem[] | undefined,
+): ProductItem[] | undefined => {
+  if (!dirtyPaymentProducts) return
+  if (!form.payments_field.products) return
+
+  const sanitisedProducts = form.payments_field.products
+    .map((cleanProductData): ProductItem | null => {
+      const dirtyProduct = dirtyPaymentProducts.find(({ data }) =>
+        (cleanProductData._id as unknown as Types.ObjectId).equals(data._id),
+      )
+      if (!dirtyProduct) return null
+
+      return {
+        ..._.pick(dirtyProduct, ['selected', 'quantity']), // only selected and quantity are allowed to be passed through
```

Comment:
any particular reason to use lodash `pick` here as opposed to just `{ ...dirtyProduct, data: cleanProductData }` or for even more safety, `{ selected: dirtyProduct.selected, quantity: dirtyProduct.quantity, data: cleanProductData }` so that we can take advantage of the type system for typechecking? Since with `pick`, the keys are not typechecked.

Reply:
> any particular reason to use lodash pick here as opposed to just { ...dirtyProduct, data: cleanProductData }

My intention was to operate as an explicit whitelist of properties that we will read from `dirtyProduct`, as opposed to accepting all the fields that are passed in. I felt that these properties are critical thus better to have them explicitly inherited rather than implicitly accepted.


> Since with pick, the keys are not typechecked.

Lodash's `pick` does have typechecking, the selected property `U` must exist as a key in object `T`.

```pick<T extends object, U extends keyof T>(object: T, ...props: Array<Many<U>>): Pick<T, U>;```